File: bindings/dart/test/subtree_test.dart
// test/subtree_test.dart
import 'dart:io';
import 'dart:typed_data';
import 'package:test/test.dart';
import 'package:wavedb/wavedb.dart';

/// Test fixture for managing database lifecycle
class SubtreeTestFixture {
  late String dbPath;
  WaveDB? db;

  void setUp() {
    final tempDir = Directory.systemTemp.createTempSync('wavedb_subtree_test_');
    dbPath = tempDir.path;
    db = WaveDB(dbPath);
  }

  void tearDown() {
    db?.close();
    db = null;
    try {
      Directory(dbPath).deleteSync(recursive: true);
    } catch (_) {}
  }
}

void main() {
  group('Subtree', () {
    late SubtreeTestFixture fixture;

    setUp(() {
      fixture = SubtreeTestFixture();
      fixture.setUp();
    });

    tearDown(() {
      fixture.tearDown();
    });

    group('open/close', () {
      test('should open and close a subtree', () {
        final subtree = fixture.db!.openSubtree('layer/graph');
        expect(subtree.isClosed, isFalse);
        expect(subtree.delimiter, equals('/'));
        subtree.close();
        expect(subtree.isClosed, isTrue);
      });

      test('should support custom delimiter', () {
        final subtree = fixture.db!.openSubtree('layer', delimiter: '.');
        expect(subtree.delimiter, equals('.'));
        subtree.close();
      });

      test('should handle double close', () {
        final subtree = fixture.db!.openSubtree('test');
        subtree.close();
        subtree.close(); // Should not throw
        expect(subtree.isClosed, isTrue);
      });
    });

    group('sync put/get/del', () {
      test('should put and get a string value', () {
        final subtree = fixture.db!.openSubtree('app');
        try {
          subtree.putSync('users/alice/name', 'Alice');
          final value = subtree.getSync('users/alice/name');
          expect(value, equals('Alice'));
        } finally {
          subtree.close();
        }
      });

      test('should put and get binary value', () {
        final subtree = fixture.db!.openSubtree('app');
        try {
          final binary = Uint8List.fromList([0x01, 0x02, 0x03, 0xFF]);
          subtree.putSync('binary/key', binary);
          final value = subtree.getSync('binary/key');
          expect(value, isA<Uint8List>());
          expect(value, equals(binary));
        } finally {
          subtree.close();
        }
      });

      test('should return null for missing keys', () {
        final subtree = fixture.db!.openSubtree('app');
        try {
          final value = subtree.getSync('missing/key');
          expect(value, isNull);
        } finally {
          subtree.close();
        }
      });

      test('should delete a key', () {
        final subtree = fixture.db!.openSubtree('app');
        try {
          subtree.putSync('key', 'value');
          subtree.delSync('key');
          final value = subtree.getSync('key');
          expect(value, isNull);
        } finally {
          subtree.close();
        }
      });

      test('should handle missing key delete', () {
        final subtree = fixture.db!.openSubtree('app');
        try {
          // Deleting non-existent key should not throw
          expect(() => subtree.delSync('missing/key'), returnsNormally);
        } finally {
          subtree.close();
        }
      });
    });

    group('cross-subtree isolation', () {
      test('subtrees with different prefixes should be isolated', () {
        final subtree1 = fixture.db!.openSubtree('namespace1');
        final subtree2 = fixture.db!.openSubtree('namespace2');
        try {
          subtree1.putSync('key', 'value1');
          subtree2.putSync('key', 'value2');

          expect(subtree1.getSync('key'), equals('value1'));
          expect(subtree2.getSync('key'), equals('value2'));
        } finally {
          subtree1.close();
          subtree2.close();
        }
      });

      test('subtree keys should not leak to main database', () {
        final subtree = fixture.db!.openSubtree('isolated');
        try {
          subtree.putSync('testkey', 'testvalue');

          // The key in the main db is prefixed: "isolated/testkey"
          // Direct access with the same bare key should not find it
          expect(fixture.db!.getSync('testkey'), isNull);
        } finally {
          subtree.close();
        }
      });
    });

    group('count', () {
      test('should count entries under the subtree prefix', () {
        final subtree = fixture.db!.openSubtree('metrics');
        try {
          expect(subtree.count(), equals(0));

          subtree.putSync('counter1', '10');
          subtree.putSync('counter2', '20');
          subtree.putSync('counter3', '30');

          expect(subtree.count(), equals(3));
        } finally {
          subtree.close();
        }
      });

      test('count should decrease after delete', () {
        final subtree = fixture.db!.openSubtree('metrics');
        try {
          subtree.putSync('a', '1');
          subtree.putSync('b', '2');
          expect(subtree.count(), equals(2));

          subtree.delSync('a');
          expect(subtree.count(), equals(1));
        } finally {
          subtree.close();
        }
      });
    });

    group('deleteSubtree (prefix delete)', () {
      test('should delete all keys under a prefix', () {
        // Put keys in the database under a known prefix
        fixture.db!.putSync('myapp/users/alice', 'Alice');
        fixture.db!.putSync('myapp/users/bob', 'Bob');
        fixture.db!.putSync('myapp/settings/theme', 'dark');

        fixture.db!.deleteSubtree('myapp');

        expect(fixture.db!.getSync('myapp/users/alice'), isNull);
        expect(fixture.db!.getSync('myapp/users/bob'), isNull);
        expect(fixture.db!.getSync('myapp/settings/theme'), isNull);
      });
    });

    group('batch', () {
      test('should execute batch operations on subtree', () {
        final subtree = fixture.db!.openSubtree('batchns');
        try {
          subtree.batchSync([
            {'type': 'put', 'key': 'key1', 'value': 'value1'},
            {'type': 'put', 'key': 'key2', 'value': 'value2'},
            {'type': 'del', 'key': 'nonexistent'},
          ]);

          expect(subtree.getSync('key1'), equals('value1'));
          expect(subtree.getSync('key2'), equals('value2'));
        } finally {
          subtree.close();
        }
      });
    });

    group('scan', () {
      test('should scan entries under a prefix within subtree', () {
        final subtree = fixture.db!.openSubtree('scan');
        try {
          subtree.putSync('users/alice', 'Alice');
          subtree.putSync('users/bob', 'Bob');
          subtree.putSync('items/x', 'Item X');

          final results = subtree.scanSyncRaw('users');
          expect(results.length, equals(2));

          final keys = results.map((e) => e['key'] as String).toList();
          // Keys returned by subtree scan have the subtree prefix stripped
          expect(keys, contains('users/alice'));
          expect(keys, contains('users/bob'));
        } finally {
          subtree.close();
        }
      });

      test('should return empty list for no matches', () {
        final subtree = fixture.db!.openSubtree('scan');
        try {
          final results = subtree.scanSyncRaw('nonexistent');
          expect(results, isEmpty);
        } finally {
          subtree.close();
        }
      });
    });

    group('snapshot', () {
      test('should snapshot the underlying database', () {
        final subtree = fixture.db!.openSubtree('snap');
        try {
          subtree.putSync('key', 'value');
          // Snapshot should not throw
          expect(() => subtree.snapshot(), returnsNormally);
        } finally {
          subtree.close();
        }
      });
    });

    group('operations on closed subtree', () {
      test('should throw on put after close', () {
        final subtree = fixture.db!.openSubtree('closed');
        subtree.close();
        expect(
          () => subtree.putSync('key', 'value'),
          throwsA(isA<WaveDBException>()),
        );
      });

      test('should throw on get after close', () {
        final subtree = fixture.db!.openSubtree('closed');
        subtree.close();
        expect(
          () => subtree.getSync('key'),
          throwsA(isA<WaveDBException>()),
        );
      });

      test('should throw on del after close', () {
        final subtree = fixture.db!.openSubtree('closed');
        subtree.close();
        expect(
          () => subtree.delSync('key'),
          throwsA(isA<WaveDBException>()),
        );
      });
    });

    group('schema collision', () {
      test('GraphLayer without subtree should fail on GraphQLLayer database', () {
        // Create a GraphQL layer on a temp database
        final tempDir = Directory.systemTemp.createTempSync('wavedb_collision_test_');
        final tempPath = tempDir.path;
        try {
          final graphql = GraphQLLayer();
          graphql.create(GraphQLLayerConfig(path: tempPath));
          graphql.parseSchema('type User { name: String }');
          graphql.close();

          // Now try to create a GraphLayer on the same database without subtree
          // This should fail with a conflict error (error code -3)
          expect(
            () => GraphLayer(tempPath),
            throwsA(isA<WaveDBException>().having((e) => e.code, 'code', 'CONFLICT')),
          );
        } finally {
          try {
            Directory(tempPath).deleteSync(recursive: true);
          } catch (_) {}
        }
      });
    });

    group('async operations', () {
      test('should put and get a value asynchronously', () async {
        final subtree = fixture.db!.openSubtree('async_test');
        await subtree.put('key1', 'value1');
        final result = await subtree.get('key1');
        expect(result, equals('value1'));
        subtree.close();
      });

      test('should return null for missing key asynchronously', () async {
        final subtree = fixture.db!.openSubtree('async_test');
        final result = await subtree.get('nonexistent');
        expect(result, isNull);
        subtree.close();
      });

      test('should delete a value asynchronously', () async {
        final subtree = fixture.db!.openSubtree('async_test');
        await subtree.put('key1', 'value1');
        final result = await subtree.get('key1');
        expect(result, equals('value1'));

        await subtree.del('key1');
        final resultAfterDelete = await subtree.get('key1');
        expect(resultAfterDelete, isNull);
        subtree.close();
      });

      test('should handle multiple async operations concurrently', () async {
        final subtree = fixture.db!.openSubtree('async_concurrent');

        // Fire multiple puts concurrently
        await Future.wait([
          subtree.put('key1', 'value1'),
          subtree.put('key2', 'value2'),
          subtree.put('key3', 'value3'),
        ]);

        // Verify all values
        expect(await subtree.get('key1'), equals('value1'));
        expect(await subtree.get('key2'), equals('value2'));
        expect(await subtree.get('key3'), equals('value3'));
        subtree.close();
      });

      test('should reject operations on closed subtree', () async {
        final subtree = fixture.db!.openSubtree('async_closed');
        await subtree.put('key1', 'value1');
        subtree.close();

        expect(
          () => subtree.put('key2', 'value2'),
          throwsA(isA<WaveDBException>()),
        );
      });
    });
  });
}

File: bindings/nodejs/test/subtree.test.js
const assert = require('assert');
const { WaveDB } = require('../lib/wavedb.js');
const { Subtree } = require('../lib/subtree.js');
const { GraphLayer } = require('../lib/graph.js');
const { GraphQLLayer } = require('../lib/graphql.js');
const fs = require('fs');
const path = require('path');

describe('Subtree', () => {
  let db;
  let testDbPath;

  beforeEach(() => {
    testDbPath = `/tmp/wavedb-subtree-test-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
    db = new WaveDB(testDbPath);
  });

  afterEach(() => {
    if (db) {
      db.close();
    }
    if (fs.existsSync(testDbPath)) {
      fs.rmSync(testDbPath, { recursive: true });
    }
  });

  describe('open/close', () => {
    it('should open and close a subtree', () => {
      const st = db.openSubtree('myns');
      assert.strictEqual(st.isClosed, false);
      assert.strictEqual(st.delimiter, '/');
      st.close();
      assert.strictEqual(st.isClosed, true);
    });

    it('should open a subtree with custom delimiter', () => {
      const st = db.openSubtree('myns', ':');
      assert.strictEqual(st.delimiter, ':');
      st.close();
    });

    it('should throw on invalid prefix', () => {
      assert.throws(() => db.openSubtree(''), /Prefix is required/);
      assert.throws(() => db.openSubtree(123), /Prefix is required/);
    });

    it('should throw on invalid delimiter', () => {
      assert.throws(() => db.openSubtree('x', ''), /Delimiter must be a single character/);
      assert.throws(() => db.openSubtree('x', 'ab'), /Delimiter must be a single character/);
    });

    it('should throw when opening subtree on closed db', () => {
      db.close();
      assert.throws(() => db.openSubtree('x'), /Database is closed/);
    });
  });

  describe('sync put/get/del', () => {
    it('should put, get, and delete values in a subtree', () => {
      const st = db.openSubtree('testns');
      st.putSync('key1', 'value1');
      assert.strictEqual(st.getSync('key1'), 'value1');
      st.delSync('key1');
      assert.strictEqual(st.getSync('key1'), null);
      st.close();
    });

    it('should handle array keys', () => {
      const st = db.openSubtree('testns');
      st.putSync(['users', 'alice', 'name'], 'Alice');
      assert.strictEqual(st.getSync(['users', 'alice', 'name']), 'Alice');
      st.close();
    });

    it('should handle binary values', () => {
      const st = db.openSubtree('testns');
      const buf = Buffer.from([0x01, 0x02, 0x03]);
      st.putSync('binkey', buf);
      const result = st.getSync('binkey');
      assert.deepStrictEqual(result, buf);
      st.close();
    });

    it('should return null for missing keys', () => {
      const st = db.openSubtree('testns');
      assert.strictEqual(st.getSync('nonexistent'), null);
      st.close();
    });

    it('should throw when operating on closed subtree', () => {
      const st = db.openSubtree('testns');
      st.close();
      assert.throws(() => st.putSync('k', 'v'), /Subtree is closed/);
      assert.throws(() => st.getSync('k'), /Subtree is closed/);
      assert.throws(() => st.delSync('k'), /Subtree is closed/);
    });
  });

  describe('cross-subtree isolation', () => {
    it('should isolate keys between subtrees with different prefixes', () => {
      const st1 = db.openSubtree('ns1');
      const st2 = db.openSubtree('ns2');

      st1.putSync('samekey', 'from-ns1');
      st2.putSync('samekey', 'from-ns2');

      assert.strictEqual(st1.getSync('samekey'), 'from-ns1');
      assert.strictEqual(st2.getSync('samekey'), 'from-ns2');

      st1.close();
      st2.close();
    });

    it('should not leak subtree keys into the main db namespace', () => {
      const st = db.openSubtree('isolate');
      st.putSync('mykey', 'myvalue');

      // The main db should see the key under the prefixed path
      const mainVal = db.getSync('isolate/mykey');
      assert.strictEqual(mainVal, 'myvalue');

      // But not under the raw key
      assert.strictEqual(db.getSync('mykey'), null);

      st.close();
    });
  });

  describe('batch', () => {
    it('should execute batch operations synchronously', () => {
      const st = db.openSubtree('batchns');
      st.batchSync([
        { type: 'put', key: 'a', value: '1' },
        { type: 'put', key: 'b', value: '2' },
        { type: 'put', key: 'c', value: '3' },
      ]);

      assert.strictEqual(st.getSync('a'), '1');
      assert.strictEqual(st.getSync('b'), '2');
      assert.strictEqual(st.getSync('c'), '3');

      st.batchSync([
        { type: 'del', key: 'b' },
      ]);

      assert.strictEqual(st.getSync('b'), null);
      st.close();
    });
  });

  describe('scan', () => {
    it('should scan keys in the subtree', () => {
      const st = db.openSubtree('scanns');
      st.putSync('users/alice', 'Alice');
      st.putSync('users/bob', 'Bob');
      st.putSync('items/x', 'X');

      const results = st.scanSyncRaw('users');
      assert.ok(results.length >= 2);

      st.close();
    });
  });

  describe('count', () => {
    it('should count entries in the subtree', () => {
      const st = db.openSubtree('countns');
      assert.strictEqual(st.count(), 0);

      st.putSync('key1', 'val1');
      st.putSync('key2', 'val2');
      st.putSync('key3', 'val3');

      assert.strictEqual(st.count(), 3);
      st.close();
    });
  });

  describe('deleteSubtree', () => {
    it('should delete all keys under a prefix', () => {
      const st = db.openSubtree('delns');
      st.putSync('key1', 'val1');
      st.putSync('key2', 'val2');
      st.close();

      db.deleteSubtree('delns');

      // Re-open same prefix, should be empty
      const st2 = db.openSubtree('delns');
      assert.strictEqual(st2.count(), 0);
      st2.close();
    });
  });

  describe('snapshot', () => {
    it('should not throw on snapshot', () => {
      const st = db.openSubtree('snapns');
      st.putSync('key1', 'val1');
      assert.doesNotThrow(() => st.snapshot());
      st.close();
    });
  });
});

describe('Subtree with Graph/GraphQL layers', () => {
  it('should expose _getPtr for cross-addon pointer passing', () => {
    const testPath = `/tmp/wavedb-subtree-ptr-test-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
    const db2 = new WaveDB(testPath);
    const st = db2.openSubtree('ptrtest');
    // _getPtr returns a BigInt (the raw C pointer value) for 64-bit safety
    const ptr = st._getPtr();
    assert.strictEqual(typeof ptr, 'bigint');
    assert.ok(ptr > 0n, 'Pointer should be a non-zero BigInt');
    st.close();
    db2.close();
    if (fs.existsSync(testPath)) {
      fs.rmSync(testPath, { recursive: true });
    }
  });
});